torch/csrc/Dtype.cpp

```diff
@@ -1,134 +1,115 @@
 #include <torch/csrc/Dtype.h>
 
-#include <cstring>
-#include <structmember.h>
 #include <torch/csrc/Exceptions.h>
-#include <torch/csrc/utils/object_ptr.h>
-#include <torch/csrc/utils/python_strings.h>
-#include <torch/csrc/utils/tensor_dtypes.h>
-#include <torch/csrc/utils/tensor_types.h>
+#include <torch/csrc/utils/pybind.h>
 
-#include <torch/csrc/Exceptions.h>
+#include <sstream>
 
-PyObject * THPDtype_New(at::ScalarType scalar_type, const std::string& name)
-{
-  AT_ASSERT(name.length() < DTYPE_NAME_LEN);
-  auto type = (PyTypeObject*)&THPDtypeType;
-  auto self = THPObjectPtr{type->tp_alloc(type, 0)};
-  if (!self) throw python_error();
-  auto self_ = reinterpret_cast<THPDtype*>(self.get());
-  self_->scalar_type = scalar_type;
-  std::strncpy(self_->name, name.c_str(), DTYPE_NAME_LEN);
-  return self.release();
-}
+namespace torch {
+using at::ScalarType;
 
-PyObject *THPDtype_is_floating_point(THPDtype *self, PyObject *noargs)
-{
-  if (at::isFloatingType(self->scalar_type)) {
-    Py_RETURN_TRUE;
-  } else {
-    Py_RETURN_FALSE;
-  }
-}
+namespace {
+#define PY_DTYPE(scalar_type, ...) { true, ScalarType::scalar_type, __VA_ARGS__ }
 
-PyObject *THPDtype_is_complex(THPDtype *self, PyObject *noargs)
-{
-  if (at::isComplexType(self->scalar_type)) {
-    Py_RETURN_TRUE;
-  } else {
-    Py_RETURN_FALSE;
-  }
+constexpr size_t NUM_DTYPES = static_cast<size_t>(ScalarType::NumOptions);
+// Must be in exactly the same order as ScalarType definitions
+constexpr PyDtype dtype_registry[NUM_DTYPES] = {
```

Comment:
actually, one benefit of the `getDtypeNames()` function that we lose here (which you delete in favor of this struct in the next commit) is that it doesn't require any duplication.

This constexpr array will fail at compile time if we're missing  a ScalarType (good), but you pointed out that the ordering needs to exactly match what's defined in `ScalarType.h`, and you had to add a static assert to check that the order in this array is kept in sync.

If someone changes the ordering in `ScalarType.h` without matching that change here they'll get a compile error. Which is okay, but it'd probably be nicer to have a single source of truth. Which we get by just having a function mapping from ScalarTypes -> names.

You could make that function constexpr though to keep the constexpr-ness of this table.

Reply:
e.g. (I haven't checked if this compiles)

```
constexpr PyDtype all_scalar_types[] = {
       AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_QINTS(PyDtype{ ScalarType::scalar_type, getDtypeNames(ScalarType::scalar_type) },)
```